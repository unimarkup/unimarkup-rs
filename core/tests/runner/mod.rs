use unimarkup_commons::test_runner::{
    self, as_snapshot::AsSnapshot, snap_test_runner::SnapTestRunner, test_file,
};
use unimarkup_core::Unimarkup;

use crate::snapshot::Snapshot;

pub fn test_block_snapshots() -> Vec<libtest_mimic::Trial> {
    let tests_path = unimarkup_commons::crate_tests_path!();

    let test_cases = test_runner::collect_tests(
        tests_path.join("spec/markup"),
        tests_path.join("spec/snapshots/"),
        "markup",
    );

    let mut test_runs = Vec::with_capacity(test_cases.len());

    for case in test_cases {
        let spec_test_name = format!("{}::spec::{}", module_path!(), case.test.name.as_str());
        let snap_test_name = format!("{}::snap::{}", module_path!(), case.test.name.as_str());

        let case2 = case.clone();
        let spec_test_run = move || {
            std::panic::catch_unwind(|| run_spec_test(case2)).map_err(|err| {
                let panic_msg = err
                    .downcast_ref::<&str>()
                    .unwrap_or(&"Panic message not available");

                format!("Test case panicked: {}", panic_msg).into()
            })
        };

        let snap_test_run = move || {
            std::panic::catch_unwind(|| run_snap_test(case)).map_err(|err| {
                let panic_msg = err
                    .downcast_ref::<&str>()
                    .unwrap_or(&"Panic message not available");

                format!("Test case panicked: {}", panic_msg).into()
            })
        };

        test_runs.push(libtest_mimic::Trial::test(spec_test_name, spec_test_run));
        test_runs.push(libtest_mimic::Trial::test(snap_test_name, snap_test_run));
    }

    test_runs
}

fn run_spec_test(case: test_runner::test_file::TestCase) {
    test_runner::spec_test::assert_um_spec(
        &case.file_name,
        &case.test,
        unimarkup_commons::config::Config::default(),
        |test: &test_file::Test, cfg| {
            let input = test.input.trim_end();
            let um = Unimarkup::parse(input, cfg);
            test_file::TestOutputs {
                html: Some(um.render_html().unwrap().to_string()),
                um: Some(test.input.clone()),
            }
        },
    );
}

fn run_snap_test(case: test_runner::test_file::TestCase) {
    let mut cfg = unimarkup_commons::config::Config::default();
    let scanner = unimarkup_commons::scanner::Scanner::default();

    let symbols = scanner.scan_str(&case.test.input);

    let mut snap_runner = SnapTestRunner::with_fn::<_, _>(&case.test.name, &symbols, |_input| {
        let um = unimarkup_core::parser::parse_unimarkup(&case.test.input, &mut cfg);

        (
            um.blocks
                .iter()
                .map(|block| Snapshot(block).as_snapshot())
                .collect(),
            &[],
        )
    })
    .with_info(format!(
        "Test '{}' from: {}",
        case.test.name, case.file_path
    ));

    if let Some(ref description) = case.test.description {
        snap_runner = snap_runner.with_description(description);
    }

    unimarkup_commons::run_snap_test!(snap_runner, &case.out_path);
}
