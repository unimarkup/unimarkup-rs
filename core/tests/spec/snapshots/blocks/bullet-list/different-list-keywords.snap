---
source: core/tests/runner/mod.rs
description: "Parse bullet list with different keywords\n"
info: "Test 'different-list-keywords' from: markup\\blocks\\bullet-list.yml"
---
BulletList(
BulletListEntry(
star list entry

BulletList(
BulletListEntry(
nested minus list entry
)
)
)
BulletListEntry(
plus list entry

BulletList(
BulletListEntry(
nested star list entry
)
)
)
)
---
With input:

* star list entry
  - nested minus list entry
+ plus list entry
  * nested star list entry


